src/imars3d/backend/workflow/validate.py

```diff
@@ -0,0 +1,86 @@
+import json
+import jsonschema
+from pathlib import Path
+from typing import Any, Dict, Union
+
+FilePath = Union[Path, str]
+
+# JSON schema. Cut be here or in its own file
+# http://json-schema.org/learn/getting-started-step-by-step.html
+SCHEMA = {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "properties": {
+        "instrument": {"type": "string"},
+        "ipts": {"type": "string"},
+        "name": {"type": "string"},
+        "workingdir": {"type": "string"},
+        "outputdir": {"type": "string"},
+        "tasks": {
+            "type": "array",
+            "minItems": 1,
+            "items": {
+                "properties": {
+                    "name": {"type": "string"},
+                    "function": {"type": "string"},
+                    "inputs": {"type": "object"},
+                    "outputs": {"type": "array"},
+                },
+                "required": ["name", "function", "inputs"],
+            },
+        },
+    },
+    "required": ["instrument", "ipts", "name", "workingdir", "outputdir", "tasks"],
+}
+
+
+def _validate_schema(json_obj: Any) -> None:
+    """Validate the data against the schema for jobs"""
+    jsonschema.validate(json_obj, schema=SCHEMA)
+
+
+def _validate_tasks(json_obj: Dict) -> None:
+    for step, task in enumerate(json_obj["tasks"]):
+        func_str = task["function"].strip()
+        if not func_str:
+            # TODO need better exception
+            raise RuntimeError(f'Step {step} specified empty "function"')
+
+
+def validates(json_str: str) -> None:
+    # verify that the string is non-empty
+    if len(json_str.strip()) == 0:
+        raise json.JSONDecodeError("Empty string", json_str, 0)
+    json_obj = json.loads(json_str)
+
+    # validation
+    _validate_schema(json_obj)
+    _validate_tasks(json_obj)
+
+
+def validate(filename: FilePath) -> None:
+    filepath = Path(filename)
+    with open(filepath, "r") as handle:
+        json_obj = json.load(handle)
+
+    # validation
+    _validate_schema(json_obj)
+    _validate_tasks(json_obj)
+
+
+class JSONValid:
+    def __init__(self, schema):
+        self._schema = schema
+
+    def __get__(self, obj, objtype=None):
+        return obj._json
+
+    def __set__(self, obj, json):
+        self._validate(json)
```

Comment:
my initial idea was for input argument `json` to be a dictionary and my `_validate(json)` do what you have implemented in `_validate_schema(json_obj)` and `_validate_tasks(json_obj)`. 

But it doesn't have to be a dictionary. In fact, `__set__` can handle `json` being a path to a file or a string containing the whole JSON, provided your validation functions return the dictionary:
```python
def __set__(self, obj, json: str):
    if json is a filepath:
        obj._json = validate(json)
    else:
        obj._json = validates(json)
```
Then later in the `Consumer` object:
```python
def __init__(self, json: str):
    self.json = json  # first validation, then implicit conversion from file to dict, or from str to dict
```